File: mocks/psak-mock/src/main/java/no/nav/pensjon/vtp/mocks/psak/PsakpselvPersonAdapter.java
package no.nav.pensjon.vtp.mocks.psak;

import no.nav.lib.pen.psakpselv.asbo.person.*;
import no.nav.pensjon.vtp.testmodell.kodeverk.Diskresjonskoder;
import no.nav.pensjon.vtp.testmodell.personopplysning.*;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.GregorianCalendar;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Optional.ofNullable;

@Component
public class PsakpselvPersonAdapter {
    private final PersonIndeks personIndeks;
    private final PersonModellRepository personModellRepository;

    public PsakpselvPersonAdapter(PersonIndeks personIndeks, PersonModellRepository personModellRepository) {
        this.personIndeks = personIndeks;
        this.personModellRepository = personModellRepository;
    }

    public Optional<ASBOPenPerson> getASBOPenPerson(String fodselsnummer) {
        return ofNullable(personModellRepository.findById(fodselsnummer))
                .flatMap(person -> personIndeks.findById(fodselsnummer)
                        .map(personopplysninger -> toASBOPerson(person, personopplysninger)));
    }

    public ASBOPenPerson toASBOPerson(final PersonModell person, final Personopplysninger personopplysninger) {
        ASBOPenPerson asboPenPerson = populateAsboPenPerson(person);
        asboPenPerson.setRelasjoner(fetchRelasjoner(personopplysninger));
        return asboPenPerson;
    }

    private static ASBOPenPerson populateAsboPenPerson(PersonModell søker) {
        ASBOPenPerson asboPenPerson = new ASBOPenPerson();
        asboPenPerson.setFodselsnummer(søker.getIdent());
        asboPenPerson.setFornavn(søker.getFornavn());
        asboPenPerson.setEtternavn(søker.getEtternavn());
        asboPenPerson.setKortnavn(søker.getFornavn() + " " + søker.getEtternavn());
        asboPenPerson.setStatus(søker.getPersonstatusFoo().getKode().name());
        asboPenPerson.setStatusKode(søker.getPersonstatusFoo().getKode().name());
        asboPenPerson.setDiskresjonskode(ofNullable(søker.getDiskresjonskode()).map(Diskresjonskoder::toString).orElse(null));
        asboPenPerson.setDodsdato(fetchDate(søker.getDødsdato()).orElse(null));
        asboPenPerson.setSivilstand(søker.getSivilstandFoo().getKode().name());
        asboPenPerson.setSivilstandDato(fetchDate(søker.getSivilstandFoo().getEndringstidspunkt())
                .orElse(fetchDate(søker.getFødselsdato())
                        .orElse(null)));
        asboPenPerson.setSprakKode(søker.getSpråk2Bokstaver());
        asboPenPerson.setSprakBeskrivelse(søker.getSpråk());
        asboPenPerson.setBostedsAdresse(søker.getAdresse(AdresseType.BOSTEDSADRESSE)
                .map(a -> (GateadresseModell)a)
                .map(PsakpselvPersonAdapter::convert)
                .orElse(null));
        asboPenPerson.setErEgenansatt(false);
        asboPenPerson.setBrukerprofil(new ASBOPenBrukerprofil());
        ASBOPenPersonUtland personUtland = new ASBOPenPersonUtland();
        personUtland.setStatsborgerKode("NOR");
        personUtland.setStatsborgerskap("NOR");
        asboPenPerson.setPersonUtland(personUtland);
        ASBOPenHistorikk historikk = createHistorikk();
        asboPenPerson.setHistorikk(historikk);

        ASBOPenUtbetalingsinformasjon utbetalingsinformasjon = new ASBOPenUtbetalingsinformasjon();
        ASBOPenNorskKonto norskKonto = new ASBOPenNorskKonto();
        norskKonto.setKontonummer("1234.45678.0232.7777");

        utbetalingsinformasjon.setNorskKonto(norskKonto);
        utbetalingsinformasjon.setUtbetalingsType("NORKNT");
        asboPenPerson.setUtbetalingsinformasjon(utbetalingsinformasjon);

        return asboPenPerson;
    }

    @NotNull
    private static ASBOPenHistorikk createHistorikk() {
        ASBOPenHistorikk historikk = new ASBOPenHistorikk();
        historikk.setBostedsadresser(null);//må være enten være populert eller null

        ASBOPenAnnenAdresseListe adresseLinjer = new ASBOPenAnnenAdresseListe();
        adresseLinjer.setAdresseLinjer(new ASBOPenAnnenAdresse[0]);
        historikk.setAdresseLinjer(adresseLinjer);

        ASBOPenHistoriskFnrListe fnrListe = new ASBOPenHistoriskFnrListe();
        fnrListe.setHistoriskeFnr(new ASBOPenHistoriskFnr[0]);
        historikk.setHistoriskeFnr(fnrListe);

        ASBOPenNavnEndringListe navnListe = new ASBOPenNavnEndringListe();
        navnListe.setNavnEndringer(new ASBOPenNavnEndring[0]);
        historikk.setNavnEndringer(navnListe);

        return historikk;
    }

    private ASBOPenRelasjonListe fetchRelasjoner(Personopplysninger personopplysninger) {
        List<ASBOPenRelasjon> relasjonList =  personopplysninger.getFamilierelasjoner().stream()
                .map(fr -> populateAsboPenRelasjon(personopplysninger, fr))
                .collect(Collectors.toList());

        ASBOPenRelasjonListe liste = new ASBOPenRelasjonListe();
        liste.setRelasjoner(relasjonList.toArray(ASBOPenRelasjon[]::new));
        return liste;
    }

    private ASBOPenRelasjon populateAsboPenRelasjon(Personopplysninger personopplysninger, FamilierelasjonModell fr) {
        ASBOPenRelasjon relasjon = new ASBOPenRelasjon();
        relasjon.setRelasjonsType(fr.getRolle().name());
        ASBOPenPerson annen = Stream.ofNullable(personopplysninger.getAnnenPart())
                .filter(p -> p.getIdent().equals(ofNullable(personModellRepository.findById(fr.getTil())).orElseThrow(() -> new RuntimeException("No person with ident " + fr.getTil())).getIdent()))
                .map(PsakpselvPersonAdapter::populateAsboPenPerson)
                .peek(p -> relasjon.setFom(fetchDate(personopplysninger.getSøker().getSivilstandFoo().getFom()).orElse(null)))
                .peek(p -> relasjon.setTom(fetchDate(personopplysninger.getSøker().getSivilstandFoo().getTom()).orElse(null)))
                .findFirst().orElseGet(() -> createShallowASBOPerson(fr));
        relasjon.setPerson(annen);

        //annen part må ha en lik relasjon også
        ASBOPenPerson shallowSoker = new ASBOPenPerson();
        shallowSoker.setFodselsnummer(personopplysninger.getSøker().getIdent());

        ASBOPenRelasjon annenPartRelasjon = new ASBOPenRelasjon();
        annenPartRelasjon.setRelasjonsType(relasjon.getRelasjonsType());
        annenPartRelasjon.setFom(relasjon.getFom());
        annenPartRelasjon.setFom(relasjon.getTom());
        annenPartRelasjon.setAdresseStatus(relasjon.getAdresseStatus());
        annenPartRelasjon.setPerson(shallowSoker);

        ASBOPenRelasjonListe liste = new ASBOPenRelasjonListe();
        liste.setRelasjoner(new ASBOPenRelasjon[]{annenPartRelasjon});
        annen.setRelasjoner(liste);

        return relasjon;
    }

    private ASBOPenPerson createShallowASBOPerson(FamilierelasjonModell familierelasjon) {
        ASBOPenPerson annen = new ASBOPenPerson();
        annen.setFodselsnummer(
                ofNullable(personModellRepository.findById(familierelasjon.getTil())).orElseThrow(() -> new RuntimeException("No person with ident " + familierelasjon.getTil())).getIdent());
        return annen;
    }

    private static Optional<GregorianCalendar> fetchDate(LocalDate date) {
        return ofNullable(date)
                .map(d -> GregorianCalendar.from(d.atStartOfDay(ZoneId.systemDefault())));
    }

    private static ASBOPenBostedsAdresse convert(GateadresseModell adresse) {
        ASBOPenBostedsAdresse asboPenBostedsAdresse = new ASBOPenBostedsAdresse();
        asboPenBostedsAdresse.setAdresseType(AdresseType.BOSTEDSADRESSE.name());
        asboPenBostedsAdresse.setBolignr(adresse.getHusnummer().toString());
        asboPenBostedsAdresse.setBoadresse1(adresse.getGatenavn() + " " + adresse.getHusnummer().toString());
        asboPenBostedsAdresse.setPostnummer(adresse.getPostnummer());
        return asboPenBostedsAdresse;
    }
}

File: mocks/psak-mock/src/main/java/no/nav/pensjon/vtp/mocks/psak/PsakpselvPersonAdapter.kt
package no.nav.pensjon.vtp.mocks.psak

import no.nav.lib.pen.psakpselv.asbo.person.*
import no.nav.pensjon.vtp.core.util.asGregorianCalendar
import no.nav.pensjon.vtp.testmodell.personopplysning.*
import no.nav.pensjon.vtp.testmodell.personopplysning.AdresseType.BOSTEDSADRESSE
import org.springframework.stereotype.Component

@Component
class PsakpselvPersonAdapter(
    private val personIndeks: PersonIndeks,
    private val personModellRepository: PersonModellRepository
) {
    fun getASBOPenPerson(fodselsnummer: String): ASBOPenPerson? {
        val personModell = personModellRepository.findById(fodselsnummer)
        val personopplysninger = personIndeks.findById(fodselsnummer)

        return if (personModell != null && personopplysninger != null) {
            toASBOPerson(personModell, personopplysninger)
        } else {
            null
        }
    }

    fun toASBOPerson(person: PersonModell, personopplysninger: Personopplysninger): ASBOPenPerson {
        return populateAsboPenPerson(person).apply {
            relasjoner = fetchRelasjoner(personopplysninger)
        }
    }

    private fun fetchRelasjoner(personopplysninger: Personopplysninger): ASBOPenRelasjonListe {
        return ASBOPenRelasjonListe().apply {
            relasjoner = personopplysninger.familierelasjoner
                .map { populateAsboPenRelasjon(personopplysninger, it) }
                .toTypedArray()
        }
    }

    private fun populateAsboPenRelasjon(
        personopplysninger: Personopplysninger,
        fr: FamilierelasjonModell
    ): ASBOPenRelasjon {
        val relasjon = ASBOPenRelasjon().apply {
            relasjonsType = fr.rolle.name
        }

        val annen = personopplysninger.annenPart
            ?.takeIf { it.ident == personModellRepository.findById(fr.til)?.ident ?: throw RuntimeException("No person with ident " + fr.til) }
            ?.let { populateAsboPenPerson(it) }
            ?.also {
                relasjon.fom = personopplysninger.søker.getSivilstandFoo().fom?.asGregorianCalendar()
                relasjon.tom = personopplysninger.søker.getSivilstandFoo().tom?.asGregorianCalendar()
            }
            ?: createShallowASBOPerson(fr)
        relasjon.person = annen

        // annen part må ha en lik relasjon også

        annen.relasjoner = ASBOPenRelasjonListe().apply {
            relasjoner = arrayOf(
                ASBOPenRelasjon().apply {
                    relasjonsType = relasjon.relasjonsType
                    fom = relasjon.fom
                    fom = relasjon.tom
                    adresseStatus = relasjon.adresseStatus
                    person = ASBOPenPerson().apply {
                        fodselsnummer = personopplysninger.søker.ident
                    }
                }
            )
        }
        return relasjon
    }

    private fun createShallowASBOPerson(familierelasjon: FamilierelasjonModell): ASBOPenPerson {
        return ASBOPenPerson().apply {
            fodselsnummer = personModellRepository.findById(familierelasjon.til)?.ident
                ?: throw RuntimeException("No person with ident " + familierelasjon.til)
        }
    }

    companion object {
        private fun populateAsboPenPerson(søker: PersonModell): ASBOPenPerson {
            return ASBOPenPerson().apply {
                fodselsnummer = søker.ident
                fornavn = søker.fornavn
                etternavn = søker.etternavn
                kortnavn = "${søker.fornavn} ${søker.etternavn}"
                status = søker.getPersonstatusFoo().kode.name
                statusKode = søker.getPersonstatusFoo().kode.name
                diskresjonskode = søker.diskresjonskode?.toString()
                dodsdato = søker.dødsdato?.asGregorianCalendar()
                sivilstand = søker.getSivilstandFoo().kode.name
                sivilstandDato = søker.getSivilstandFoo().endringstidspunkt?.asGregorianCalendar() ?: søker.fødselsdato?.asGregorianCalendar()
                sprakKode = søker.getSpråk2Bokstaver()
                sprakBeskrivelse = søker.språk
                bostedsAdresse = søker.getAdresse(BOSTEDSADRESSE)?.asASBOPenBostedsAdresseOrNull()
                erEgenansatt = false
                brukerprofil = ASBOPenBrukerprofil()
                personUtland = ASBOPenPersonUtland().apply {
                    statsborgerKode = "NOR"
                    statsborgerskap = "NOR"
                }
                historikk = createHistorikk()

                utbetalingsinformasjon = ASBOPenUtbetalingsinformasjon().apply {
                    utbetalingsType = "NORKNT"
                    norskKonto = ASBOPenNorskKonto().apply {
                        kontonummer = "1234.45678.0232.7777"
                    }
                }
            }
        }

        private fun createHistorikk(): ASBOPenHistorikk {
            return ASBOPenHistorikk().apply {
                bostedsadresser = null // må være enten være populert eller null
                adresseLinjer = ASBOPenAnnenAdresseListe().apply { adresseLinjer = arrayOfNulls(0) }
                historiskeFnr = ASBOPenHistoriskFnrListe().apply { historiskeFnr = arrayOfNulls(0) }
                navnEndringer = ASBOPenNavnEndringListe().apply { navnEndringer = arrayOfNulls(0) }
            }
        }
    }
}

fun AdresseModell.asASBOPenBostedsAdresseOrNull(): ASBOPenBostedsAdresse? {
    return if (this is GateadresseModell) {
        ASBOPenBostedsAdresse().apply {
            adresseType = BOSTEDSADRESSE.name
            bolignr = husnummer.toString()
            boadresse1 = "$gatenavn $husnummer"
            postnummer = this.postnummer
        }
    } else {
        null
    }
}
